linux/devdoc/completion_port_linux_requirements.md

```diff
@@ -0,0 +1,187 @@
+
+If `completion_port` is `NULL`, `completion_port_add` shall return a non-NULL value.
+
+If `ingress_data` is `NULL`, `completion_port_add` shall return a non-NULL value.
+
+`completion_port_add` shall allocate a `EPOLL_THREAD_DATA` object to store thread data.
+
+`completion_port_add` shall add the `EPOLL_THREAD_DATA` object to a list for later removal
+
+`completion_port_add` shall add the socket in the epoll system by calling `epoll_ctl` with `EPOLL_CTL_MOD` along with the `COMPLETION_PORT_INGRESS_DATA` `epoll_op` variable.
+
+If the `epoll_ctl` call fails with `ENOENT`, `completion_port_add` shall call `epoll_ctl` again with `EPOLL_CTL_ADD`.
+
+On success, `completion_port_add` shall return 0.
+
+If any error occurs, `completion_port_add` shall fail and return a non-zero value.
+
+### completion_port_remove
+
+```C
+MOCKABLE_FUNCTION(, void, completion_port_remove, COMPLETION_PORT_HANDLE, completion_port, SOCKET_HANDLE, socket);
+```
+
+If `completion_port` is `NULL`, `completion_port_remove` shall return.
+
+If `socket` is `INVALID_SOCKET`, `completion_port_remove` shall return.
+
+`completion_port_remove` shall remove the underlying socket from the epoll by calling `epoll_ctl` with `EPOLL_CTL_DEL`.
+
+### epoll_worker_func
+
+```c
+static int epoll_worker_func(void* parameter)
+```
+
+If `parameter` is `NULL`, `epoll_worker_func` shall do nothing.
+
+`epoll_worker_func` shall call `epoll_wait` to wait for an epoll event to become signaled with a timeout of 2 Seconds.
+
+If the events value contains `EPOLLRDHUP` (hang up), `epoll_worker_func` shall the following:
+
+- `epoll_worker_func` shall receive the `EPOLL_THREAD_DATA` value from the ptr variable from the `epoll_event` data ptr.
+
+- Then call the `event_callback` callback with the `event_callback_ctx` and `COMPLETION_PORT_RESULT_ABANDONED`.
+
+- Then remove the `EPOLL_THREAD_DATA` from the list and free the object
+
+If the events value contains `EPOLLIN` (recv), `epoll_worker_func` shall do the following:
+
+- `epoll_worker_func` shall receive the `EPOLL_THREAD_DATA` value from the ptr variable from the `epoll_event` data ptr.
+
+- Then `epoll_worker_func` shall call `recv` and do the following:
+
+- If the recv size < 0, then:
+
+  - If `errno` is either `EAGAIN` or `EWOULDBLOCK`, then `epoll_worker_func` shall break and call `event_callback` callback with the `event_callback_context` with `COMPLETION_PORT_RESULT_OK`
+
+  - If `errno` is `ECONNRESET`, then `epoll_worker_func` shall call the `event_callback` callback with the `event_callback_context` and `COMPLETION_PORT_RESULT_ABANDONED`
+
+- If the recv size equal 0, then `epoll_worker_func` shall call `event_callback` callback with the `event_callback_context` and `COMPLETION_PORT_RESULT_OK`
+
+- If the recv size > 0, then if recv size < buffer length is less than `epoll_worker_func` shall issue another recv, otherwise we shall call `event_callback` callback with the `event_callback_context` with `COMPLETION_PORT_RESULT_OK`
```

Comment:
> > 0, then if recv size < buffer  [](http://example.com/codeflow?start=18&length=33)

I think the fact that we have 2 ifs in the same spec points to "let's break it down into at least 2 sentences